### PR TITLE
ptyfwd: some tweaks to terminal handling

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

File: src/run/run.c [hunk 1]
```diff
@@ -1584,6 +1592,24 @@ static int acquire_invocation_id(sd_bus *bus, const char *unit, sd_id128_t *ret)
...
 }
 
+static void set_window_title(PTYForward *f) {
+        _cleanup_free_ char *hn = NULL, *cl = NULL, *dot = NULL;
+        assert(f);
+
+        if (!arg_host)
+                (void) gethostname_strict(&hn);
+
+        cl = strv_join(arg_cmdline, " ");
```
Comment (YHNdnzj) [inline src/run/run.c:1602]: I think we should just give up if this is empty. The new window title carries no useful information if so after all...

File: src/run/run.c [hunk 2]
```diff
@@ -1584,6 +1592,24 @@ static int acquire_invocation_id(sd_bus *bus, const char *unit, sd_id128_t *ret)
...
+
+        if (emoji_enabled())
+                dot = strjoin(special_glyph(privileged_execution() ? SPECIAL_GLYPH_RED_CIRCLE : SPECIAL_GLYPH_YELLOW_CIRCLE), " ");
+
+        if (arg_host || hn)
+                (void) pty_forward_set_titlef(f, "%s%s on %s", strempty(dot), strempty(cl), arg_host ?: hn);
+        else
+                (void) pty_forward_set_titlef(f, "%s%s", strempty(dot), strempty(cl));
```
Comment (YHNdnzj) [inline src/run/run.c]: (especially in this case)

**Comment by poettering** [comment]
New version pushed. PTAL.

File: src/run/run.c [hunk 3]
```diff
@@ -1584,6 +1592,26 @@ static int acquire_invocation_id(sd_bus *bus, const char *unit, sd_id128_t *ret)
...
+        if (!cl)
+                return (void) log_oom();
+
+        if (emoji_enabled())
+                dot = strjoin(special_glyph(privileged_execution() ? SPECIAL_GLYPH_RED_CIRCLE : SPECIAL_GLYPH_YELLOW_CIRCLE), " ");
+
+        if (arg_host || hn)
+                (void) pty_forward_set_titlef(f, "%s%s on %s", strempty(dot), strempty(cl), arg_host ?: hn);
+        else
+                (void) pty_forward_set_titlef(f, "%s%s", strempty(dot), strempty(cl));
```
Comment (YHNdnzj) [inline src/run/run.c]: Then, there's no need to use `strempty` for `cl`.

**Review by YHNdnzj — COMMENTED** [review]
Superficially looks good, but I have poor knowledge about ANSI escape codes though. So leaving the approve button to others.

**Comment by mrc0mmand** [comment]
This does some weird things when nspawn is used in scripts:

```
[    6.880565] testsuite-06.sh[367]: + NSPAWN_ARGS=(systemd-nspawn -q --volatile=yes --directory=/ --bind-ro=/etc --inaccessible=/etc/machine-id)
[    6.881732] testsuite-06.sh[414]: ++ systemd-nspawn -q --volatile=yes --directory=/ --bind-ro=/etc --inaccessible=/etc/machine-id cat /proc/self/attr/current
[    6.881983] testsuite-06.sh[416]: ++ tr -d '\0'
[    6.889143] systemd[1]: run-u6.service was skipped because of an unmet condition check (ConditionSecurity=!selinux).
[    6.934169] systemd[1]: Created slice machine.slice.
[    6.934420] systemd[1]: var-lib-machines.mount was skipped because of an unmet condition check (ConditionPathExists=/var/lib/machines.raw).
[    6.934949] systemd[1]: Starting systemd-machined.service...
[    6.971263] systemd[1]: Started systemd-machined.service.
[    6.978270] systemd-machined[420]: New machine H.
[    6.980772] systemd[1]: Couldn't stat device /dev/net/tun: No such file or directory
[    6.981685] systemd[1]: Started machine-H.scope.
[    7.110838] systemd[1]: machine-H.scope: Deactivated successfully.
[    7.111029] systemd-machined[420]: Machine H terminated.
[    7.132128] testsuite-06.sh[367]: + [[ ontainer H on H^Gsystem_u:system_r:systemd_nspawn_t != \s\y\s\t\e\m\_\u\:\o\b\j\e\c\t\_\r\:\s\y\s\c\t\l\_\k\e\r\n\e\l\_\t ]]
[    7.153968] testsuite-06.sh[424]: ++ systemd-nspawn -q --volatile=yes --directory=/ --bind-ro=/etc --inaccessible=/etc/machine-id --selinux-context=system_u:object_r:sysctl_kernel_t cat /proc/self/attr/current
[    7.154225] testsuite-06.sh[425]: ++ tr -d '\0'
[    7.272173] systemd-machined[420]: New machine H.
[    7.296587] systemd[1]: Couldn't stat device /dev/net/tun: No such file or directory
[    7.297453] systemd[1]: Started machine-H.scope.
[    7.391410] systemd[1]: machine-H.scope: Deactivated successfully.
[    7.391491] systemd-machined[420]: Machine H terminated.
[    7.425947] testsuite-06.sh[367]: + [[ ontainer H on H^Gsystem_u:object_r:sysctl_kernel_t == \s\y\s\t\e\m\_\u\:\o\b\j\e\c\t\_\r\:\s\y\s\c\t\l\_\k\e\r\n\e\l\_\t ]]
[    7.560992] systemd[1]: run-systemd-nspawn-unix\x2dexport-H.mount: Deactivated successfully.
[    7.563672] systemd[1]: testsuite-06.service: Main process exited, code=exited, status=1/FAILURE
[    7.563901] systemd[1]: testsuite-06.service: Failed with result 'exit-code'.
[    7.564443] systemd[1]: Failed to start testsuite-06.service.
[    7.575427] systemd[1]: Reached target testsuite.target.
```

**Review by bluca — APPROVED** [review]

**Pull request merged**